src/librustc_mir/transform/elaborate_drops.rs

```diff
@@ -389,10 +459,30 @@ impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
                 }
             };
 
-            on_all_drop_children_bits(self.tcx, self.mir, self.env, path, |child| {
+            // Create drop flags for each child path P of the thing we
+            // are dropping, if that path P may or may not be live
+            // (this is the case where we need runtime state to tell). So for example:
+            //
+            // ```
+            // fn foo(x: (Vec<u32>, u32) {
+            //     if rand() {
+            //         mem::drop(x.0);
+            //         mem::drop(x.1);
+            //     }
+            // }
+            // ```
+            //
+            // Here, `x` would be dropped, and there are two child
+            // paths of `x` (`x.0`, `x.1`).  `x.1` has a `Copy` type
+            // so we would ignore that. `x.0` may be live and may be
+            // dead, so we would add a drop flag for that.
+            //
+            // NDM -- if there was e.g. a move of `x.0` and also
+            // `x.0.0`, would we create flags for both?
```

Comment:
in the original fragments code, my mental model was that you would have one flag for `x.0.0`, and another for `all_remaining_unmentioned_children_of(x.0)`. (One reason I had taken this approach is that the amount of drop flag state is bounded by the number of paths referenced in the code itself, rather than being dependent on the number of fields in the structure.)

Then a move of `x.0` would modify both of those two flags.

I don't yet know if this code does the same thing. I mostly mention it in passing here.

Reply:
Ah yes, this code is just using the same move paths data that is built up by https://github.com/rust-lang/rust/blob/master/src/librustc_mir/dataflow/move_paths/builder.rs (the `on_all_drop_children_bits` traversal is itself defined in https://github.com/rust-lang/rust/blob/master/src/librustc_mir/dataflow/drop_flag_effects.rs but that is unlikely to answer these sorts of questions; one really needs to consult the move path construction code for the answers.

Reply:
@nikomatsakis 

I think so. Both `x.0` and `x.0.0` are possibly live and possibly dead, so we create a flag at each level of the recursion.

This is rare enough that it does not pose a perf problem in practice.